pkg/resource/deploy/step_generator.go

```diff
@@ -72,8 +72,51 @@ func (sg *stepGenerator) isTargetedUpdate() bool {
 	return sg.updateTargetsOpt != nil || sg.replaceTargetsOpt != nil
 }
 
-func (sg *stepGenerator) isTargetedForUpdate(urn resource.URN) bool {
-	return sg.updateTargetsOpt == nil || sg.updateTargetsOpt[urn]
+// isTargetedForUpdate returns if urn is target for update. The function accommodates
+// `--target-dependents`. `targetDependentsForUpdate` should probably be called if this function
+// returns true.
+func (sg *stepGenerator) isTargetedForUpdate(
+	urn, parent resource.URN, provider string, dependencies []resource.URN) bool {
+	if sg.updateTargetsOpt == nil || sg.updateTargetsOpt[urn] {
+		return true
+	}
+	if provider != "" {
+		res, err := providers.ParseReference(provider)
+		contract.AssertNoError(err)
+		dependencies = append(dependencies, res.URN())
```

Comment:
I would avoid modifying the input slice like this. It shouldn't be visible to callers--and won't be if the slice gets copied during the `append`--but instead, I'd rather we manually check `sg.updateTargetsOpt` for the URN.